recipes/gettext/all/test_package/conanfile.py

```diff
@@ -17,35 +21,38 @@ def requirements(self):
          self.requires(self.tested_reference_str)
 
     def build_requirements(self):
-        self.build_requires(self.tested_reference_str)
-        self.build_requires("automake/1.16.4")
-        if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
-            self.build_requires("msys2/cci.latest")
-
-    @contextlib.contextmanager
-    def _build_context(self):
-        if self.settings.compiler == "Visual Studio":
-            with tools.vcvars(self):
-                env = {
-                    "CC": "cl -nologo",
-                    "LD": "link -nologo",
-                }
-                with tools.environment_append(env):
-                    yield
-        else:
-            yield
+        self.tool_requires(self.tested_reference_str)
+        self.tool_requires("automake/1.16.5")
+        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", check_type=str):
+            self.tool_requires("msys2/cci.latest")
+
+    def layout(self):
+        basic_layout(self, src_folder="src")
+        
+    def generate(self):
+        at = AutotoolsToolchain(self)
+        at.generate()
+
+        if is_msvc(self):
+            env = Environment()
+            env.define("CC", "cl -nologo")
+            env.define("LD", "link -nologo")
+            env.vars(self).save_script("conanbuild_libsmacker_msvc")
+
+        runenv = VirtualRunEnv(self)
+        runenv.generate()
+
+        buildenv = VirtualBuildEnv(self)
+        buildenv.generate()
```

Comment:
```suggestion
        buildenv = VirtualBuildEnv(self)
        buildenv.generate()

        at = AutotoolsToolchain(self)
        at.generate()

        if is_msvc(self):
            env = Environment()
            env.define("CC", "cl -nologo")
            env.define("LD", "link -nologo")
            env.vars(self).save_script("conanbuild_libsmacker_msvc")

        runenv = VirtualRunEnv(self)
        runenv.generate()
```

VirtualBuildEnv first, always, because some generators like AutotoolsToolchain also manipulate env vars, and while VirtualBuildEnv is stupid, other generators may sanitize or override env vars in a smarter way.